cmd/enterprise-portal/internal/samsm2m/samsm2m.go

```diff
@@ -0,0 +1,99 @@
+package samsm2m
+
+import (
+	"context"
+	"net/http"
+
+	"connectrpc.com/connect"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	otelcodes "go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/sourcegraph/log"
+
+	sams "github.com/sourcegraph/sourcegraph-accounts-sdk-go"
+	"github.com/sourcegraph/sourcegraph-accounts-sdk-go/scopes"
+
+	"github.com/sourcegraph/sourcegraph/cmd/enterprise-portal/internal/connectutil"
+	"github.com/sourcegraph/sourcegraph/internal/authbearer"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// EnterprisePortalScope returns the Enterprise Portal service scope for the
+// given permission and action.
+func EnterprisePortalScope(permission scopes.Permission, action scopes.Action) scopes.Scope {
+	return scopes.ToScope(scopes.ServiceEnterprisePortal, permission, action)
+}
+
+var tracer = otel.GetTracerProvider().Tracer("telemetry-gateway/samsm2m")
+
+type TokenIntrospector interface {
+	IntrospectToken(ctx context.Context, token string) (*sams.IntrospectTokenResponse, error)
+}
+
+type Request interface {
+	Header() http.Header
+}
+
+// RequireScope ensures the request context has a valid SAMS M2M token
+// with requiredScope. It returns a ConnectRPC status error suitable to be
+// returned directly from a ConnectRPC implementation.
+//
+// See: go/sams-m2m
+func RequireScope(ctx context.Context, logger log.Logger, tokens TokenIntrospector, requiredScope scopes.Scope, req Request) (err error) {
+	logger = logger.Scoped("samsm2m")
+
+	var span trace.Span
+	ctx, span = tracer.Start(ctx, "RequireScope")
+	defer func() {
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(otelcodes.Error, "check failed")
+		}
+		span.End()
+	}()
+
+	var token string
+	if v := req.Header().Values("authorization"); len(v) == 1 && v[0] != "" {
```

Comment:
```suggestion
	if v := req.Header().Get("authorization"); v != "" {
```

You can do this if you want

Reply:
I do want to make sure there's only one, explicitly 😋

Reply:
Does it matter in the perspective of this package 😂 As long as the first one is correct

Reply:
Hm, I think it does matter; IMO multiple authorization is an error case. Michael and I ran into a similar issue with cookies recently, where multiple cookies with the same name would cause weird and difficult-to-debug errors, because an error case (multiple cookies with the same name) was being silently accepted as correct (but it wasn't)

Reply:
>where multiple cookies with the same name would cause weird and difficult-to-debug errors

I'd say that's a feature for cookies? Unless you mean something else... browser should only send the one with the narrowest scope.

Reply:
Sure it could be a feature, but not one that's correct, or helpful 😁

Reply:
Resolving for now IMO, in favour of being careful and explicit